### PR TITLE
Table rebalance status

### PR DESCRIPTION
**Pull request by antwigambrah**
Description:
This PR  adds an api for checking table rebalance status

This PR will solve the following issue --https://github.com/apache/pinot/issues/9558

cc @Jackie-Jiang

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java [hunk 1]
```diff
@@ -383,6 +383,19 @@ public RebalanceResult rebalance(TableConfig tableConfig, Configuration rebalanc
     }
   }
 
+  public RebalanceResult.Status rebalanceStatus(String tableNameWithType, IdealState idealState,
+      ExternalView externalView) {
+
+    if (idealState == null) {
+      throw new IllegalStateException("ideal state is empty");
+    }
+    if (externalView != null && isExternalViewConverged(tableNameWithType, externalView.getRecord().getMapFields(),
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:392]: Do we want to consider the situation where external view never converges with what ongoing rebalancer expects or external view ends up having segments in ERROR state ? 

Looks like we will continue returning IN_PROGRESS in such scenarios and that could be misleading ? Probably also depends on what is the call to action for the user/admin who uses this API...

Comment (antwigambrah) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java]: HI @siddharthteotia, sorry I updated the above code before reading this comment.Currently, the external view returns  false if it is not converged yet and logs an error if the external view ends up having segments in error .I believe it is best to return the status FAILED instead of logging. Also, would it be ok to return NO_OP if the external view is still null when the api is called .cc @Jackie-Jiang

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java [hunk 2]
```diff
@@ -383,6 +383,23 @@ public RebalanceResult rebalance(TableConfig tableConfig, Configuration rebalanc
...
+    if (idealState == null) {
+      throw new IllegalStateException("ideal state is empty");
+    }
+    if (externalView == null) {
+      return RebalanceResult.Status.NO_OP;
+    }
+
+    if (isExternalViewConverged(tableNameWithType, externalView.getRecord().getMapFields(),
+        idealState.getRecord().getMapFields(), true)) {
+      return RebalanceResult.Status.DONE;
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:398]: I am not sure if it makes sense / helpful enough to sort of implement a stateless API to get status of rebalance task -- which was triggered separately through some other API and is being done asynchronously on potentially a different controller that this API call comes on

Relying on the fact that `ExternalView` is not the same as `IdealState` to indicate Rebalance `IN_PROGRESS` could be misleading. 

Segment assignment changes ideal state as soon as segment lands on controller and up until server successfully downloads the segment etc, `ExternalView` does not get updated. But this API will return rebalance is in `IN_PROGRESS` even though rebalance may have already finished or could have errored out etc.

So not sure if this is going to be super helpful / accurate.

Comment (siddharthteotia) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:398]: cc @antwigambrah @Jackie-Jiang thoughts ?

Comment (antwigambrah) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:398]: HI, @siddharthteotia  Yes it will be quite misleading. Checking for IN_PROGRESS seems tricky here. 

The possible way I can think for now through testing I have done on TableRebalancerClusterStatelessTest is to check on the external view from being null to being allocated segments assignments. 

For DONE status, I will run the rebalance in dry mode and retrieve the new assignments from the result and use is checking if the extervalViewConverged . Kindly let me know you thoughts on it . cc @Jackie-Jiang

**Review by Jackie-Jiang — COMMENTED** [review]
The API added in this PR basically checks if the table's EV matches IS. This itself cannot represent the status of rebalance because rebalance can be done in multiple steps, and in the middle EV can match IS. But this is also a quite useful endpoint for status tracking purpose.
I'd suggest changing the API to return whether the table has EV converged, and if not, list down the segments that do not have the same state in EV and IS.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 3]
```diff
@@ -687,6 +688,26 @@ public RebalanceResult rebalance(
...
   }
 
+  @GET
+  @Path("/tables/{tableName}/rebalance/status")
+  @ApiOperation(value = "Table rebalance status", notes = "Get segments without the same state in EV and IS ")
+  public String getTableRebalanceStatus(
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
+      @ApiParam(value = "OFFLINE|REALTIME", required = true) @QueryParam("type") String tableTypeStr) {
+    try {
+        String tableNameWithType = constructTableNameWithType(tableName, tableTypeStr);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:698]: (code format) Please reformat the changes with [Pinot Style](https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup#setup-ide)

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 4]
```diff
@@ -687,6 +688,26 @@ public RebalanceResult rebalance(
     }
   }
 
+  @GET
+  @Path("/tables/{tableName}/rebalance/status")
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:692]: Let's make the API more specific, and it might be better to put it under the segment rest API: `/segments/{tableName}/externalViewMismatch` and returns a map from the mismatch segment to its ideal state and external view

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 5]
```diff
@@ -3082,6 +3083,17 @@ public RebalanceResult rebalanceTable(String tableNameWithType, Configuration re
     return new TableRebalancer(_helixZkManager).rebalance(tableConfig, rebalanceConfig);
   }
 
+  public Map<String, MapDifference.ValueDifference<Map<String, String>>> rebalanceTableStatus(String tableNameWithType)
+      throws TableNotFoundException {
+    TableConfig tableConfig = getTableConfig(tableNameWithType);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:3088]: No need to read the table config. We should check both IS and EV exist

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java [hunk 6]
```diff
@@ -383,6 +385,25 @@ public RebalanceResult rebalance(TableConfig tableConfig, Configuration rebalanc
     }
   }
 
+  public Map<String, MapDifference.ValueDifference<Map<String, String>>> rebalanceStatus(String tableNameWithType,
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:388]: Don't put this into the `TableRebalancer`. We can directly have it in `PinotHelixResourceManager`

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 7]
```diff
@@ -733,6 +754,7 @@ private String constructTableNameWithType(String tableName, String tableTypeStr)
     try {
       tableType = TableType.valueOf(tableTypeStr.toUpperCase());
     } catch (Exception e) {
+
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:757]: (nit) Please revert some unnecessary empty line changes in this PR

**Comment by antwigambrah** [comment]
Hi @Jackie-Jiang   , Please review with the new update

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 8]
```diff
@@ -732,6 +732,7 @@ private String constructTableNameWithType(String tableName, String tableTypeStr)
...
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:735]: (minor) Revert this change

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java [hunk 9]
```diff
@@ -421,13 +421,13 @@ private InstancePartitions getInstancePartitions(TableConfig tableConfig,
     if (InstanceAssignmentConfigUtils.allowInstanceAssignment(tableConfig, instancePartitionsType)) {
       if (reassignInstances) {
         String rawTableName = TableNameBuilder.extractRawTableName(tableNameWithType);
-        boolean hasPreConfiguredInstancePartitions = TableConfigUtils.hasPreConfiguredInstancePartitions(tableConfig,
-            instancePartitionsType);
+        boolean hasPreConfiguredInstancePartitions =
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java:424]: (minor) Revert the changes in this file since it is not related

File: pinot-controller/src/test/java/org/apache/pinot/controller/api/PinotSegmentRestletResourceTest.java [hunk 10]
```diff
@@ -195,6 +195,32 @@ private void checkCrcRequest(Map<String, SegmentMetadata> metadataTable, int exp
...
+    String createTableUrl = TEST_INSTANCE.getControllerRequestURLBuilder().forTableCreate();
+    TableConfigBuilder offlineBuilder = new TableConfigBuilder(TableType.OFFLINE);
+
+    TableConfig testTableConfig = offlineBuilder.setTableName(tableName).build();
+
+    // Create the table
+    ControllerTest.sendPostRequest(createTableUrl, testTableConfig.toJsonString());
+
+    // Rebalance table
+    ControllerTest.sendPostRequest(
```
Comment (Jackie-Jiang) [inline pinot-controller/src/test/java/org/apache/pinot/controller/api/PinotSegmentRestletResourceTest.java:211]: There is no segment in this table, so both rebalance and segment mismatch will be no-op. Let's try to generate a non-empty IS and EV with some diff, and verify the behavior

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java [hunk 11]
```diff
@@ -1072,4 +1075,34 @@ private SuccessResponse updateZKTimeIntervalInternal(String tableNameWithType) {
...
+      String tableNameWithType = TableNameBuilder.forType(TableType.valueOf(tableTypeStr)).tableNameWithType(tableName);
+
+      if (!_pinotHelixResourceManager.hasTable(tableNameWithType)) {
+        throw new TableNotFoundException(String.format("Table=%s not found", tableName));
+      }
+
+      Map<String, MapDifference.ValueDifference<Map<String, String>>> view =
+          _pinotHelixResourceManager.getExternalViewSegementMismatch(tableNameWithType);
+      ObjectNode data = JsonUtils.newObjectNode();
+      data.put("segments", view.toString());
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java:1101]: I feel we might want the response to be the following format (`segmentName` -> `EV/IS` -> different `instanceState`): 
```
{
  "segment0": {
    "IdealState": {
      "server0": "ONLINE",
      "server1": "ONLINE"
    },
    "externalView": {
      "server1": "ERROR",
      "server2": "ONLINE"
    }
  },
  "segment1": {
    "IdealState": {
      "server1": "ONLINE"
    },
    "externalView": {
      "server1": "ERROR"
    }
  }
}
```

**Review by Jackie-Jiang — COMMENTED** [review]
The overall direction looks good

**Comment by antwigambrah** [comment]
Hi @Jackie-Jiang  , an update has been provided for this pr. Kindly review

**Comment by Jackie-Jiang** [comment]
Decided to move forward with #10359
